src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/PinnedLineWidget.java

```diff
@@ -0,0 +1,226 @@
+ *
+ */
+package org.rstudio.studio.client.workbench.views.source.editors.text;
+
+import org.rstudio.core.client.HandlerRegistrations;
+import org.rstudio.studio.client.workbench.views.source.editors.text.ace.Anchor;
+import org.rstudio.studio.client.workbench.views.source.editors.text.ace.LineWidget;
+import org.rstudio.studio.client.workbench.views.source.editors.text.ace.Position;
+import org.rstudio.studio.client.workbench.views.source.editors.text.events.FoldChangeEvent;
+
+import com.google.gwt.core.client.JavaScriptObject;
+import com.google.gwt.user.client.Command;
+import com.google.gwt.user.client.ui.Widget;
+
+/*
+ * A pinned line widget binds an Ace LineWidget to a GWT widget. It provides
+ * two features needed by most usages of LineWidgets:
+ * 
+ * - The LineWidget is "pinned" to its original line; a pair of anchors detects
+ *   when the widget has been shifted by editing, and restores it to its
+ *   original position.
+ * - A notification is emitted to the host when Ace removes the LineWidget.
+ */
+public class PinnedLineWidget
+             implements FoldChangeEvent.Handler
+{
+   public interface Host
+   {
+      void onLineWidgetRemoved(LineWidget widget);
+   }
+
+   public PinnedLineWidget(String type, DocDisplay display, Widget widget, 
+         int row, JavaScriptObject data, Host host)
+   {
+      display_ = display;
+      widget_ = widget;
+      host_ = host;
+      
+      lastWidgetRow_ = row;
+      moving_ = false;
+      shiftPending_ = false;
+
+      lineWidget_ = LineWidget.create(type, row, widget_.getElement(), data);
+      lineWidget_.setFixedWidth(true); 
+      display_.addLineWidget(lineWidget_);
+
+      // the Ace line widget manage emits a 'changeFold' event when a line
+      // widget is destroyed; this is our only signal that it's been
+      // removed, so when it happens, we need check to see if the widget
+      // has been removed
+      registrations_ = new HandlerRegistrations(
+         display_.addFoldChangeHandler(this));
+
+      startAnchor_ = display_.createAnchor(Position.create(row, 0));
+
+      createEndAnchor();
+   }
+   
+   public boolean moving()
+   {
+      return moving_;
+   }
+   
+   public int getRow()
+   {
+      if (shiftPending_)
+         return lastWidgetRow_;
+      return lineWidget_.getRow();
+   }
+   
+   public void detach()
+   {
+      detachAnchors();
+      registrations_.removeHandler();
+      display_.removeLineWidget(lineWidget_);
+   }
+   
+   public LineWidget getLineWidget()
+   {
+      return lineWidget_;
+   }
+   
+   // Event handlers ----------------------------------------------------------
+
+   @Override
+   public void onFoldChange(FoldChangeEvent event)
+   {
+      // the FoldChangeEvent is fired at the moment that Ace removes the line
+      // widget, but before our anchors are updated. set a flag so we know to
+      // check for an attached widget next time.
+      checkForRemove_ = true;
```

Comment:
Unfortunately not, because the removal check needs to run *after* the fold change but *before* the event loop (`syncEndAnchor` below is called prior to the event loop).